### PR TITLE
Extensibility: support the PluginDocumentSettingPanel slot in the site editor

### PR DESCRIPTION
**Pull request by youknowriad**
Related #52632 

## What?

This PR is going to set a precent about how we merge extensibility APIs between post and site editors, so we need to review and test properly.

The idea is that all the slots that exit in either post or site editors can be shared through the `editor` package. That way, a plugin registering a fill can work both in the post editor and site editor (assuming the plugin loads its JS in both pages).

## How?

- Moves the PluginDocumentSettingPanel from `@wordpress/edit-post` to. `@wordpress/editor`
- Adds the slot to the site editor as well.

## Testing Instructions

You should be able to paste something like the following in both post and site editors and have the panel show up in the sidebar of pages.

```js
var el = React.createElement;
var __ = wp.i18n.__;
var registerPlugin = wp.plugins.registerPlugin;
var PluginDocumentSettingPanel = wp.editor.PluginDocumentSettingPanel;

function MyDocumentSettingPlugin() {
	return el(
		PluginDocumentSettingPanel,
		{
			className: 'my-document-setting-plugin',
			title: 'My Panel',
			name: 'my-panel',
		},
		__( 'My Document Setting Panel' )
	);
}

registerPlugin( 'my-document-setting-plugin', {
	render: MyDocumentSettingPlugin,
} );
```

### CODE REVIEW

**Comment by Mamaduka** [comment]
I'll try to do extensive testing later today. I think I've dozens of similar snippets that use `PluginDocumentSettingPanel` to test things quickly.

Side note: @ryanwelcher and I chatted during WCAsia about posibility of the new API for registering the document settings panels.

A basic API mockup:

```js
/**
 * Registers a document settings panel for Featured Image.
 */
registerDocumentSetting( 'core/featured-image', {
	// Label displayed in Preferences modal.
	label: 'Featured Image',

	// The position in the document settings sidebar. Similar to core's admin menu pages.
	position: 5,

	// Optional. The scope for the document settings. We can also use `site` and `post`.
	scope: [ 'edit-site', 'edit-post' ],

	// A React component to render the settings.
	edit: () => {}
} )
```

**Comment by youknowriad** [comment]
@Mamaduka Interesting though, I've been thinking about this as well and had similar ideas. I think there's a potential to go even further. Instead of having to register "document settings", we would register "fields":

```
registerField( 'featured-image', 'post', {
	label: 'Featured Image',
        getValue({ item }) { return item.featured_image; }
        type: "image"
        
        // Potentially override the field type render if needed.
        render
} )
```

- The goal being that this would both register the field in data views and also in the inspector/quick edit...
- The goal is also for the API to be mostly serializable, without the "render" function you can register this in the backend. if you want something specific about the "edit" form, you can override the "render" in JS.

That said, my thinking at the moment is that we're still a bit early there and while we have "fields" as internal API for now, I think we'll probably make a lot of changes (some breaking) as we advance in dataviews, forms and the tasks related to the admin redesign.

File: packages/edit-post/src/index.js [hunk 1]
```diff
@@ -161,7 +162,7 @@ export function reinitializeEditor() {
 }
 
 export { default as PluginBlockSettingsMenuItem } from './components/block-settings-menu/plugin-block-settings-menu-item';
-export { default as PluginDocumentSettingPanel } from './components/sidebar/plugin-document-setting-panel';
+export { PluginDocumentSettingPanel };
```
Comment (fabiankaegy) [inline packages/edit-post/src/index.js:165]: Should we wrap this in a deprecated call to let extenders know about the change so they can change their implementations? 

I don't think we would need a removal timeframe for now. But from what I can tell here it doesn't make sense to use the version exported through there going forward. 

It's definitely important that it is here though to not break anything :)

Comment (youknowriad) [reply packages/edit-post/src/index.js:165]: Yes, I wondering about it. Should we deprecate now or give it some time (maybe move more slots and check feedback)

**Comment by Mamaduka** [comment]
@youknowriad, the "fields" API looks more flexible, especially when we consider new concepts being introduced for admin redesign.

Do you think there's room for both APIs? I'm guessing the "Document" and "Block" sidebars will remain even after the redesign.

**Comment by youknowriad** [comment]
> Do you think there's room for both APIs? I'm guessing the "Document" and "Block" sidebars will remain even after the redesign.

yes, I think the current APIs will remain but I'm not sure that we'd need an additional `registerDocumentSetting` that kind of sits in the middle between a "fields" API and a the slot.

**Comment by gziolo** [comment]
> The idea is that all the slots that exit in either post or site editors can be shared through the editor package. That way, a plugin registering a fill can work both in the post editor and site editor (assuming the plugin loads its JS in both pages).

It makes perfect sense to move all plugin components to `@wordpress/editor` package to allow sharing them between the post and site editor. The way it's proposed in this PR should be a good way to transition so people can continue importing the component from `wp.editPost`, but it will load the component from `wp.editor`.

That said, do we have some examples of plugins using the `PluginDocumentSettingPanel` to validate if all of them make sense when rendered in the site editor? I'm also not entirely sure which type of slots it is. What I'm trying to see we should ensure that the existing fills rendered in the site editor integrate correctly there and potentially offer a way for devs to clarify the intent.

**Comment by fabiankaegy** [comment]
One additional question that comes up for me is how we best handle some of the API inconsistencies between the two editors. 

Something I often run into is how I can access the current post id / the post type being edited. The way that is done is, as far as I'm aware, different. 

So this change will have a potential to cause the panels to be broken in the site editor.

**Comment by Mamaduka** [comment]
> So this change will have a potential to cause the panels to be broken in the site editor.

That's a possibility primarily if the fills rely on `edit-post` components or data store.

**Comment by youknowriad** [comment]
> So this change will have a potential to cause the panels to be broken in the site editor.

Adding the plugins to the site editor requires explicitly loading the JS in the site editor. So in other words, there's no breakage to be expected here.

> Something I often run into is how I can access the current post id / the post type being edited. The way that is done is, as far as I'm aware, different.

I think we've made huge improvements here lately, both editors rely on `@wordpress/editor` these days and you can get the current entity... from there. But yes, if you rely on `edit-post` it won't work.

**Comment by Mamaduka** [comment]
> Adding the plugins to the site editor requires explicitly loading the JS in the site editor. So in other words, there's no breakage to be expected here.

If a script is enqueued using `enqueue_block_editor_assets`, it will be loaded in both editors. There's no special handling needed. I'll put together a testing plugin.

**Comment by fabiankaegy** [comment]
> I think we've made huge improvements here lately, both editors rely on @wordpress/editor these days and you can get the current entity

Would be really cool to get an up to date docs page or blog post about all this :) I see the question of how to access data in the same way across the site and post editor pop up all the time :)

**Review by aaronrobertshaw — COMMENTED** [review]
FWIW I've given this a test in its current state.

- Pasting the example snippet into each editor results in the expected panel display as advertied
- Testing with a plugin using a single `enqueue_block_editor_assets` action adds the panel to both editors

So as George and Fabian flagged, it does seem to open the door for potential breakage depending on what existing plugins rely on e.g. `edit-post`.

**Comment by Mamaduka** [comment]
Based on my test, the chances of breakage are pretty low, especially if people migrate away from deprecated selectors when using WP 6.5. Unfortunately, that rarely happens; in reality, devs don't have time, or plugins have to support multiple WP versions.

Here's my example plugin that triggers an error in the site editor. Luckily, the error is contained by the plugin boundary. Repo: https://github.com/Mamaduka/test-plugin-document-setting-panel 

## Screenshot
CleanShot 2024-03-20 at 10 40 46

**Comment by youknowriad** [comment]
> Here's my example plugin that triggers an error in the site editor. Luckily, the error is contained by the plugin boundary.

Oh this is cool :P

**Comment by youknowriad** [comment]
So it feels to me like we should move forward, we're still very early in release cycle, and assess feedback received before next WP beta.

**Comment by fabiankaegy** [comment]
> So it feels to me like we should move forward, we're still very early in release cycle, and assess feedback received before next WP beta.

Agree with that. Lets get it in now so we have time to test & get feedback :)

**Review by fabiankaegy — APPROVED** [review]

**Comment by markhowellsmead** [comment]
There are two issues related to this change, as it relates to the post and page editors.

1. The [developer documentation](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-document-setting-panel/) for projects in which the Gutenberg plugin _isn't_ in use is wrong for the current version of WordPress Core.
2. Changing the source `PluginDocumentSettingPanel` from @wordpress/edit-post to @wordpress/editor is a major breaking change and should definitely be handled with deprecation notices.

These points relate to use of WordPress 6.5.3 without the Gutenberg plugin.

**Comment by fabiankaegy** [comment]
@markhowellsmead What issue are you seeing in 6.5.3? 

I just tested and `wp.editPost.PluginDocumentSettingPanel` still exists as expected? 

Also in the Gutenberg plugin the code still works just with a deprecation notice. So this should not break anything https://github.com/WordPress/gutenberg/blob/8d7c8fddfe86fffe203eef4ffe7a9653c005989c/packages/edit-post/src/deprecated.js#L39-L45

**Comment by markhowellsmead** [comment]
Apologies, the JS error was prompted by custom code in our project. 
The documentation issue is now under https://github.com/WordPress/gutenberg/issues/61657.

**Comment by juanmaguitar** [comment]
> Would be really cool to get an up-to-date docs page or blog post about all this :) I see the question of how to access data in the same way across the site and post editor pop up all the time :)

@fabiankaegy, would this blog post be similar to the following suggested idea for a Developer Blog post: https://github.com/WordPress/developer-blog-content/discussions/53? If not, would you mind opening a new [Topic Idea](https://github.com/WordPress/developer-blog-content/discussions/categories/topic-ideas) with the specific content you'd like to be covered in a Developer Blog post? I'd be up for writing such a post 🙂

**Pull request merged**